Process pure ordering changes with windowing operators

kgyrtkirk opened the pull request:
* adds a new query build path: `DruidQuery#toScanAndSortQuery` which:
   * builds a `ScanQuery` without considering the current `ordering`
   * builds an operator to execute the sort
* fixes a `null` string to `"null"` literal string conversion in the frame serializer [code](https://github.com/apache/druid/blob/c5fa649ea57b50e1b3c8861da17dc684b6143b47/processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultColumnSelectorFactoryMaker.java#L108-L112)
* fixes some `DrillWindowQueryTest` cases
* fix NPE in `NaiveSortOperator` in case there was no input
* enables back `CoreRules.AGGREGATE_REMOVE`
* adds a `processing` level `OffsetLimit` class and uses that instead of just the `limit` in the `rac` parts
* earlier window expressions on top of a subquery with an offset may have ignored the offset
```sql
  SELECT
    FLOOR(__time TO DAY) t,
    COUNT(1) OVER ()
  FROM ( select * from foo offset 3 ) f
```

somu-imply commented on processing/src/main/java/org/apache/druid/query/operator/OffsetLimit.java:
```diff
@@ -0,0 +1,139 @@
+import org.apache.druid.error.DruidException;
+import java.util.Objects;
+
+public class OffsetLimit
```
Also can we call this a OffsetLimitOperator ?

kgyrtkirk replied:
yes; we should consolidate the 2 classes - I was not able to use it and move it right away; it would have been a refactor step which I wanted to avoid as it would touch some places of the planner as well - so I would rather put it into a separate PR; in which only that's happening.

Since we have an `Operator` interface; I think it might be a bit confusing to add that keyword to this class's name - I think after we will only have one it might be already better.

somu-imply replied:
Since this is a release blocker, let's not try to refactor everything here and fix as a followup

somu-imply reviewed (commented):
Thanks for the fix. I checked out the branch and ran a set of queries. Things look good. There was one test case that failed earlier before this change.
```
with t AS (
SELECT	
		RANK() OVER (PARTITION BY m2 ORDER BY m2 ASC) 
			AS ranking,
		COUNT(m1) as trend_score
FROM foo
GROUP BY m2,m1 LIMIT 10
)
select ranking, trend_score from t ORDER BY trend_score 
```
Previously this used to fail with 
```
Query could not be planned. A possible reason is [SQL query requires order by non-time column [[trend_score ASC]], which is not supported.]
```
But now throws a NPE. Although this is under window functions and need not be investigated as a part of this PR. I'll open an issue to track it.

I have some minor comments. Rest LGTM !

kgyrtkirk commented on sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:
```diff
@@ -1473,13 +1481,78 @@ private WindowOperatorQuery toWindowQuery()
+    }
+
+    // Reject cases which would sort the datasource directly
+    if (dataSource != DruidOuterQueryRel.DUMMY_DATA_SOURCE && dataSource.isConcrete()) {
```
I intentionally wanted to check for the reference of that object - so that it identifies that it was called from [this point in DruidOuterQuery](https://github.com/apache/druid/blob/60c2ad597ab44610298e94987582443368863269/sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidOuterQueryRel.java#L123)

later this should probably revised - as it's already a bit odd that `OuterRel` passes a `DataSource` which is `concreate` - meanwhile its known to be a subquery

kgyrtkirk commented on processing/src/test/java/org/apache/druid/query/operator/NaiveSortOperatorTest.java:
```diff
@@ -0,0 +1,66 @@
+  }
+
+  @Test
+  public void testSort()
```
yeah; was thinking about adding it - but thinked that it might be covered elsewhere :D 
doesn't hurt to add a test here as well...

kgyrtkirk commented:
> But now throws a NPE. Although this is under window functions and need not be investigated as a part of this PR. I'll open an issue to track it.

it was a an issue caused by:
* ScanQuery [not returning types](https://github.com/apache/druid/blob/60c2ad597ab44610298e94987582443368863269/processing/src/main/java/org/apache/druid/query/scan/ScanQueryQueryToolChest.java#L200) for columns not being mapped by a VC
* so later on some COMPLEX column type was created as a [last restort](https://github.com/apache/druid/blob/60c2ad597ab44610298e94987582443368863269/processing/src/main/java/org/apache/druid/query/rowsandcols/ArrayListRowsAndColumns.java#L150) with a `null` inner type...
* which was dereferenced during a lookup in [by a ConcurrentHashMap](https://github.com/apache/druid/blob/60c2ad597ab44610298e94987582443368863269/processing/src/main/java/org/apache/druid/segment/column/TypeStrategies.java#L56)

I've fixed it in this PR as the underlying datasource did know the correct types; it was just not forwarded correctly - but we can remove it if that's preffered 

I've added your query to `CalciteQueryTest` :)

abhishekagarwal87 commented on sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:
```diff
@@ -1473,13 +1481,78 @@ private WindowOperatorQuery toWindowQuery()
+  }
+
   /**
    * Return this query as a Scan query, or null if this query is not compatible with Scan.
```
could you add a few more details here as to how considerSorting parameter is used?

abhishekagarwal87 commented on sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:
```diff
@@ -1473,13 +1481,78 @@ private WindowOperatorQuery toWindowQuery()
+          .stream().map(OrderByColumnSpec::getDimension)
+          .collect(Collectors.toList());
+      plannerContext.setPlanningError(
+          "SQL query requires order by non-time column [%s], which is not supported.",
```
```suggestion
          "SQL query requires ordering a table by non-time column [%s], which is not supported.",
```

abhishekagarwal87 commented on sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:
```diff
@@ -1473,13 +1481,78 @@ private WindowOperatorQuery toWindowQuery()
+    }
+
+    QueryDataSource newDataSource = new QueryDataSource(scan);
+    ArrayList<ColumnWithDirection> sortColumns = getColumnWithDriectionsFromOrderBys(sorting.getOrderBys());
```
```suggestion
    ArrayList<ColumnWithDirection> sortColumns = getColumnWithDirectionsFromOrderBys(sorting.getOrderBys());
```

kgyrtkirk commented on processing/src/main/java/org/apache/druid/query/operator/NaiveSortOperator.java:
```diff
@@ -33,11 +34,11 @@
 {
   private final Operator child;
-  private final ArrayList<ColumnWithDirection> sortColumns;
+  private final List<ColumnWithDirection> sortColumns;
```
* I think `ArrayList` have spread to a lot of places when these classes were created
* its harder to test with `ArrayList` types on the interfaces...I didn't wanted to go down further; I think wrapping stuff into the `ArrayList` when the sorter is created should be rare enough
* intellij checks also encourages to use `Collections.singletonList` and `emptyList` which are not `ArrayLists` either....

would you like to take an alternate approach?

abhishekagarwal87 replied:
No. makes sense though that makes me wonder why not just do it for NaiveSorter and its implementations too. So you don't have to create an ArrayList at line 61

kgyrtkirk commented on processing/src/main/java/org/apache/druid/query/operator/OffsetLimit.java:
```diff
@@ -0,0 +1,139 @@
+    if (hasLimit()) {
+      long toIndex = limit + offset;
+      if (limit > Long.MAX_VALUE - offset) {
+        throw DruidException.defensive(
+            "Cannot compute toIndex due to overflow [%s]",
+            this);
```
we have some places where `Long.MAX_VALUE` is substituted - I'm now feel like its even safer to not add this throw... or at least not now - I'll get back to it in a later refactor

kgyrtkirk commented on processing/src/main/java/org/apache/druid/query/rowsandcols/LazilyDecoratedRowsAndColumns.java:
```diff
@@ -238,7 +239,8 @@ private Pair<byte[], RowSignature> materializeStorageAdapter(StorageAdapter as)
       }
 
-      int theLimit = limit == -1 ? Integer.MAX_VALUE : limit;
+      long offsetRemaining = limit.getOffset();
```
totally agree :) 
renamed to `remainingRowsToSkip` and `remainingRowsToFetch`

clintropolis commented on sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:
```diff
@@ -1474,12 +1487,93 @@ private WindowOperatorQuery toWindowQuery()
+      List<String> orderByColumnNames = sorting.getOrderBys()
+          .stream().map(OrderByColumnSpec::getDimension)
+          .collect(Collectors.toList());
+      plannerContext.setPlanningError(
+          "SQL query requires ordering a table by non-time column [%s], which is not supported.",
+          orderByColumnNames);
```
super nitpick, but much of codebase puts trailing `)` on newlines so stuff is symmetrical ⚖️  (noticed this in lots of other places too)
```suggestion
      plannerContext.setPlanningError(
          "SQL query requires ordering a table by non-time column [%s], which is not supported.",
          orderByColumnNames
      );
```
not a blocker by any means, just throwing this out there since i find it more aesthetically pleasing 😅

kgyrtkirk replied:
I noticed it as well - however didn't digged into it before; I think the eclipse formatter of the project is outdated - I had to fix a few issues already (because checkstyle followed different rules).

regarding this one: it was off as well; but I don't yet see a way to enable formatting to (1) but was able to set only (2)
```
format_1(123,
    23123,
    3213,
    2321321
);

format_2(
    123,
    23123,
    3213,
    2321321
);
```
as a matter of fact I don't know if `format_1` would be the preffered or not.
I've set `format_2`; as putting the closing `)` on the same line have caused quite a few annoying mistakes for me in the last couple days...

more-or-less related: we could enforce these formatting things (see [this PR](https://github.com/apache/druid/pull/14937) ); but I didn't wanted to do it for the whole project before we agree on the rules we should follow

kgyrtkirk commented on processing/src/test/java/org/apache/druid/query/operator/NaivePartitioningOperatorFactoryTest.java:
```diff
@@ -0,0 +1,52 @@
+  @Test
+  public void testEquals()
+  {
+    new EqualsTester()
```
didn't seen that sorry - its not easier at all to use it... but its nice to `declare` equalityGroup-s :D

would you like me to remove them?

kgyrtkirk commented on processing/src/main/java/org/apache/druid/query/scan/ScanQueryQueryToolChest.java:
```diff
@@ -207,6 +208,19 @@ public RowSignature resultArraySignature(final ScanQuery query)
     }
   }
 
+  private ColumnType getDataSoruceColumnType(DataSource dataSource, String columnName)
```
jeez; I make so many typos lately that I should 
probably there are some plugins to warn about issues like this :D
I see [this](https://github.com/check-spelling/check-spelling) ; but one which could be integrated into the build would be probably even better...

fixed it

kgyrtkirk commented on sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidOuterQueryRel.java:
```diff
@@ -47,7 +47,7 @@
 public class DruidOuterQueryRel extends DruidRel<DruidOuterQueryRel>
 {
-  private static final TableDataSource DUMMY_DATA_SOURCE = new TableDataSource("__subquery__");
+  public static final TableDataSource DUMMY_DATA_SOURCE = new TableDataSource("__subquery__");
```
yes - totally agree that is not the best like this...

I think its rather odd to pass this constant [here](https://github.com/apache/druid/blob/737947754dfed23e649d980e2fc71b33ecb6e479/sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidOuterQueryRel.java#L123) as its being used to show the internal parts that this outer query has a "TableDataSource"...

I think we should probably
* remove that `?:`:
* change the `DUMMY_QUERY_DATA_SOURCE` to not look like a `ScanQuery` - instead some opaque `query`; so that it can't interfere with things
* possibly reconsider the `instanceof` at [DruidQuery#computeQuery](https://github.com/apache/druid/blob/737947754dfed23e649d980e2fc71b33ecb6e479/sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java#L981)  as when there is no `window` because of that `?:` it essentially never becomes true (during early stages of planning) because it passes a non-query datasource ....

...but I can try the `isConcreate()` impl approach for now - but if that breaks stuff; I would probably go back to the current reference checking approach - and submit the above changes in a followup PR

kgyrtkirk commented:
@somu-imply, @abhishekagarwal87, @clintropolis could you please take another look?

abhishekagarwal87 commented on sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidOuterQueryRel.java:
```diff
@@ -47,7 +47,14 @@
+  private static final TableDataSource DUMMY_DATA_SOURCE = new TableDataSource("__subquery__")
+  {
+    @Override
+    public boolean isConcrete()
```
should be helpful to have some comments associated with this override.

abhishekagarwal87 reviewed (approved):
Nit comments that need not block the PR.

abhishekagarwal87 commented:
@kgyrtkirk - You will also need to change the docs here https://druid.apache.org/docs/latest/querying/scan-query.
